File: examples/lookup.swift
import Foundation

struct Feed {
    let name: String
    let baseScore: UInt8
    let confidence: UInt8
    let flagsMask: UInt32
    let categoriesMask: UInt8
    let ipv4Starts: [UInt64]
    let ipv4Ends: [UInt64]
}

struct Blocklist {
    let timestamp: UInt32
    let flags: [String]
    let categories: [String]
    let feeds: [Feed]
}

class Reader {
    let data: Data
    var pos: Int = 0

    init(_ data: Data) { self.data = data }

    func readU8() -> UInt8 {
        let v = data[pos]; pos += 1; return v
    }

    func readU16() -> UInt16 {
        let v = data[pos..<pos+2].withUnsafeBytes {
            $0.load(as: UInt16.self)
        }
        pos += 2; return UInt16(littleEndian: v)
    }

    func readU32() -> UInt32 {
        let v = data[pos..<pos+4].withUnsafeBytes {
            $0.load(as: UInt32.self)
        }
        pos += 4; return UInt32(littleEndian: v)
    }

    func readStr() -> String {
        let len = Int(readU8())
        let s = String(data: data[pos..<pos+len], encoding: .utf8)!
        pos += len; return s
    }

    func readVarint() -> UInt64 {
        var result: UInt64 = 0; var shift: UInt64 = 0
        while true {
            let b = readU8()
            result |= UInt64(b & 0x7F) << shift
            if b & 0x80 == 0 { return result }
            shift += 7
        }
    }
}

func load(_ path: String) -> Blocklist {
    let data = try! Data(contentsOf: URL(fileURLWithPath: path))
    let r = Reader(data)

    assert(String(data: data[0..<4], encoding: .ascii) == "IPBL")
    r.pos = 4
    assert(r.readU8() == 2)
    let timestamp = r.readU32()

    let flags = (0..<r.readU8()).map { _ in r.readStr() }
    let cats = (0..<r.readU8()).map { _ in r.readStr() }
    let feedCount = r.readU16()

    var feeds: [Feed] = []
    for _ in 0..<feedCount {
        let name = r.readStr()
        let bs = r.readU8(), co = r.readU8()
        let fm = r.readU32(), cm = r.readU8()
        let rc = r.readU32()

        var v4s: [UInt64] = [], v4e: [UInt64] = []
        var current: UInt64 = 0
        for _ in 0..<rc {
            current &+= r.readVarint()
            let size = r.readVarint()
            let end = current &+ size
            if end <= 0xFFFFFFFF {
                v4s.append(current); v4e.append(end)
            }
        }
        feeds.append(Feed(
            name: name, baseScore: bs, confidence: co,
            flagsMask: fm, categoriesMask: cm,
            ipv4Starts: v4s, ipv4Ends: v4e
        ))
    }

    return Blocklist(
        timestamp: timestamp, flags: flags,
        categories: cats, feeds: feeds
    )
}

func bisectRight(_ arr: [UInt64], _ target: UInt64) -> Int {
    var lo = 0, hi = arr.count
    while lo < hi {
        let mid = lo + (hi - lo) / 2
        if arr[mid] <= target { lo = mid + 1 } else { hi = mid }
    }
    return lo
}

func ipv4ToInt(_ s: String) -> UInt64? {
    let parts = s.split(separator: ".").compactMap {
        UInt64($0)
    }
    guard parts.count == 4 else { return nil }
    return (parts[0] << 24) | (parts[1] << 16)
         | (parts[2] << 8) | parts[3]
}

let args = CommandLine.arguments
guard args.count >= 2 else {
    print("Usage: lookup <ip> [<ip> ...]"); exit(1)
}

let bl = load("blocklist.bin")

for ip in args[1...] {
    guard let target = ipv4ToInt(ip) else {
        print("\(ip): invalid/unsupported IP"); continue
    }

    var found = false
    for feed in bl.feeds {
        let idx = bisectRight(feed.ipv4Starts, target) - 1
        if idx >= 0 && target <= feed.ipv4Ends[idx] {
            let score = Double(feed.baseScore) / 200.0
                      * Double(feed.confidence) / 200.0
            var parts = [feed.name, String(format: "score=%.2f", score)]
            let fl = bl.flags.enumerated()
                .filter { feed.flagsMask & (1 << $0.offset) != 0 }
                .map(\.element)
            if !fl.isEmpty { parts.append("flags=\(fl.joined(separator: ","))") }
            let ca = bl.categories.enumerated()
                .filter { feed.categoriesMask & (1 << $0.offset) != 0 }
                .map(\.element)
            if !ca.isEmpty { parts.append("cats=\(ca.joined(separator: ","))") }
            print("\(ip): \(parts.joined(separator: " | "))")
            found = true
        }
    }

    if !found { print("\(ip): no matches") }
}

File: examples/lookup.ts
import { readFileSync } from "fs";
import { isIPv4 } from "net";

interface Feed {
    name: string;
    baseScore: number;
    confidence: number;
    flagsMask: number;
    categoriesMask: number;
    ipv4Starts: number[];
    ipv4Ends: number[];
    flags: string[];
    categories: string[];
}

interface Blocklist {
    timestamp: number;
    feeds: Feed[];
}

function load(path = "blocklist.bin"): Blocklist {
    const buf = readFileSync(path);
    let pos = 0;

    if (buf.toString("ascii", 0, 4) !== "IPBL") throw new Error("bad magic");
    pos = 4;
    if (buf[pos++] !== 2) throw new Error("bad version");
    const timestamp = buf.readUInt32LE(pos); pos += 4;

    const readStr = (): string => {
        const len = buf[pos++];
        const s = buf.toString("utf8", pos, pos + len);
        pos += len; return s;
    };

    const readVarint = (): bigint => {
        let result = 0n, shift = 0n;
        while (true) {
            const b = buf[pos++];
            result |= BigInt(b & 0x7F) << shift;
            if (!(b & 0x80)) return result;
            shift += 7n;
        }
    };

    const fc = buf[pos++];
    const flagTable = Array.from({ length: fc }, readStr);
    const cc = buf[pos++];
    const catTable = Array.from({ length: cc }, readStr);

    const feedCount = buf.readUInt16LE(pos); pos += 2;
    const feeds: Feed[] = [];

    for (let i = 0; i < feedCount; i++) {
        const name = readStr();
        const baseScore = buf[pos++];
        const confidence = buf[pos++];
        const flagsMask = buf.readUInt32LE(pos); pos += 4;
        const categoriesMask = buf[pos++];
        const rc = buf.readUInt32LE(pos); pos += 4;

        const ipv4Starts: number[] = [];
        const ipv4Ends: number[] = [];
        let current = 0n;

        for (let r = 0; r < rc; r++) {
            current += readVarint();
            const end = current + readVarint();
            if (end <= 0xFFFFFFFFn) {
                ipv4Starts.push(Number(current));
                ipv4Ends.push(Number(end));
            }
        }

        const flags = flagTable.filter((_, j) => flagsMask & (1 << j));
        const cats = catTable.filter((_, j) => categoriesMask & (1 << j));

        feeds.push({
            name, baseScore, confidence,
            flagsMask, categoriesMask,
            ipv4Starts, ipv4Ends,
            flags, categories: cats,
        });
    }

    return { timestamp, feeds };
}

function bisectRight(arr: number[], target: number): number {
    let lo = 0, hi = arr.length;
    while (lo < hi) {
        const mid = (lo + hi) >>> 1;
        if (arr[mid] <= target) lo = mid + 1;
        else hi = mid;
    }
    return lo;
}

function ipv4ToInt(ip: string): number {
    const p = ip.split(".");
    return ((+p[0] << 24) | (+p[1] << 16) | (+p[2] << 8) | +p[3]) >>> 0;
}

const args = process.argv.slice(2);
if (!args.length) {
    process.stderr.write("Usage: ts-node lookup.ts <ip> [<ip> ...]\n");
    process.exit(1);
}

const bl = load();

for (const ip of args) {
    if (!isIPv4(ip)) {
        console.log(`${ip}: invalid/unsupported IP`);
        continue;
    }

    const target = ipv4ToInt(ip);
    let found = false;

    for (const f of bl.feeds) {
        const idx = bisectRight(f.ipv4Starts, target) - 1;
        if (idx >= 0 && target <= f.ipv4Ends[idx]) {
            const score = (f.baseScore / 200) * (f.confidence / 200);
            const parts = [f.name, `score=${score.toFixed(2)}`];
            if (f.flags.length) parts.push(`flags=${f.flags.join(",")}`);
            if (f.categories.length)
                parts.push(`cats=${f.categories.join(",")}`);
            console.log(`${ip}: ${parts.join(" | ")}`);
            found = true;
        }
    }

    if (!found) console.log(`${ip}: no matches`);
}
